File: Python/Leetcode Medium/238ProductOfArrayExceptSelf.py
#https://leetcode.com/problems/product-of-array-except-self/description/
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        #1, 2, 3, 4
        #1, 1, 2, 6
        #24, 12, 4, 1
        
        ret = [1 for __ in range(len(nums))]
        tempLeft = 1
        for i in range(len(nums)-1):
            tempLeft *= nums[i]
            ret[i+1] *= tempLeft
        tempRight = 1
        for i in range(len(nums)-1):
            tempRight *= nums[len(nums)-1-i]
            ret[len(nums)-2-i] *= tempRight
        return ret
    
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        #1, 2, 3, 4
        #1, 2, 6, 24
        #24, 24, 12, 4
        left = []
        right = deque([])
        tempLeft = 1
        for num in nums:
            tempLeft *= num
            left.append(tempLeft)
        tempRight = 1
        for i in range(len(nums)):
            tempRight *= nums[len(nums)-1-i]
            right.appendleft(tempRight)
        right = list(right)
        ret = []
        for i in range(len(nums)):
            if i == 0:
                ret.append(right[1])
                continue
            if i == len(nums)-1:
                ret.append(left[len(nums)-2])
                continue
            ret.append(left[i-1] * right[i+1])
        return ret

File: C++/Leetcode Medium/238ProductOfArrayExceptSelf.cpp
#include <vector>
using namespace std;
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> res(nums.size(), 1);
        int forward = 1;
        int backward = 1;
        for (int i = 0; i < nums.size(); ++i) {
            res[i] *= forward;
            res[res.size()-1-i] *= backward;
            forward *= nums[i];
            backward *= nums[nums.size()-1-i];
        }
        return res;        
    }
};

#include <vector>
using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        //1, 2, 3, 4
        //1, 1, 2, 6
        //24, 12, 4, 1
        vector<int> forward(nums.size(), 1);
        vector<int> backward(nums.size(), 1);
        for (int i = 1; i < nums.size(); ++i) {
            forward[i] = forward[i-1] * nums[i-1];
            backward[nums.size()-1-i] = backward[nums.size()-1-i+1] * nums[nums.size()-1-i+1];
        }
        vector<int> res;
        for (int i = 0; i < forward.size(); ++i) {
            res.push_back(forward[i] * backward[i]);
        }
        return res;
    }
};